File: ch15/ex15-3.py
# 15.3 배열에동작하는 확장 함수 작성

import array
avg(array.array('d',[1,2,3]))
# 2.0
import numpy
avg(numpy.array([1.0,2.0,3.0]))
# 2.0
avg([1,2,3])
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: 'list' does not support the buffer interface
avg(b'Hello')
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: 'list' does not support the buffer interface
a = numpy.array([[1.,2.,3.],[4.,5.,6.]])
avg(a[:,2])
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# ValueError: ndarray is not contiguous
sample.avg(a)
# Traceback (most recent call last):
#   File "<stdin>", line 1, in <module>
# TypeError: Expected a 1-dimensional array
sample.avg(a[0])
# 2.0



# 토론

# typedef struct bufferinfo {
#     void *buf;                 /* 버퍼 메모리 포인터 */
#     PyObject *obj;             /* 소유자 파이썬 객체 */
#     Py_ssize_t len;            /* 전체 크기(byte) */
#     Py_ssize_t itemsize;       /* 아이템 하나의 크기(byte) */
#     int readonly;              /* 읽기 전용 플래그 */
#     int ndim;                  /* 차원 수 */
#     char *format;              /* 아이템 하나의 구조체 코드 */
#     Py_ssize_t *shape;         /* 차원 배열 */
#     Py_ssize_t *strides;       /* strides 배열 */
#     Py_ssize_t *suboffsets;    /* suboffsets 배열 */
# } Py_buffer;

File: ch15/ex15-3.c
// 15.3 배열에동작하는 확장 함수 작성

/* double avg(double *, int) 호출 */
static PyObject *py_avg(PyObject *self, PyObject *args) {
  PyObject *bufobj;
  Py_buffer view;
  double result;
  /* 전달 받은 파이썬 객체 얻기 */
  if (!PyArg_ParseTuple(args, "0", &bufobj)) {
    return NULL;
  }

  /* 여기서 버퍼 정보를 얻는다. */
  if (PyObject_GetBuffer(bufobj, &view,
      PyBUF_ANY_CONTIGUOUS | PYy_BUF_FORMAT) == -1) {
    return NULL;
  }

  if (view.ndim != 1) {
    PyErr_SetString(PyExc_TypeError, "Expected a 1-dimensional array");
    PyBuffer_Release(&view);
    return NULL;
  }

  /* 배열 속 아이템 확인 */
  if (strcmp(view.format,"d") != 0) {
    PyErr_SetString(PyExc_TypeError, "Expected an array of doubles");
    PyBuffer_Release(&view);
    return NULL;
  }

  /* 로우 버퍼와 크기를 C 함수에 전달 */
  result = avg(view.buf, view.shape[0]);

  /* 버퍼와 작업을 마쳤음을 알린다. */
  PyBuffer_Release(&view);
  return Py_BuildValue("d", result);
}
